openssl/src/cms.rs

```diff
@@ -80,4 +81,51 @@ impl CmsContentInfo {
             Ok(CmsContentInfo::from_ptr(cms))
         }
     }
+
+    /// Given a signing cert `signcert`, private key `pkey`, an optional certificate stack `certs`,
+    /// data `data` and flags `flags`, create a CmsContentInfo struct.
+    ///
+    /// OpenSSL documentation at [`CMS_sign`]
+    ///
+    /// [`CMS_sign`]: https://www.openssl.org/docs/manmaster/man3/CMS_sign.html
+    pub fn sign<T: HasPrivate>(
+        signcert: &X509,
+        pkey: &PKeyRef<T>,
+        certs: Option<&Stack<X509>>,
+        data: &[u8],
+        flags: u32,
+    ) -> Result<CmsContentInfo, ErrorStack> {
+        unsafe {
+            let signcert = signcert.as_ptr();
+            let pkey = pkey.as_ptr();
+            let data_bio = MemBioSlice::new(data)?;
+            let cms = cvt_p(ffi::CMS_sign(
+                signcert,
+                pkey,
+                certs.unwrap_or(&Stack::<X509>::new()?).as_ptr(),
+                data_bio.as_ptr(),
+                flags,
+            ))?;
+
+            Ok(CmsContentInfo::from_ptr(cms))
+        }
+    }
+
+    /// Serializes this CmsContentInfo using DER.
+    ///
+    /// OpenSSL documentation at [`i2d_CMS_ContentInfo`]
+    ///
+    /// [`i2d_CMS_ContentInfo`]: https://www.openssl.org/docs/man1.0.2/crypto/i2d_CMS_ContentInfo.html
+    pub fn to_der(&mut self) -> Result<Vec<u8>, ErrorStack> {
+        unsafe {
+            let size = ffi::i2d_CMS_ContentInfo(self.as_ptr(), ptr::null_mut());
+            let mut der = vec![0u8; size as usize];
+
+            let raw_ptr = Box::into_raw(Box::new(der.as_mut_ptr()));
```

Comment:
This leaks a pointer to a pointer.

```rust
ffi::i2d_CMS_ContentInfo(self.as_ptr(), der.as_mut_ptr());
```

Reply:
I had no idea that taking a reference to a pointer would be equivalent. TIL.